lib/thrift/include/thrift/transport/TSSLSocket.h

```diff
@@ -0,0 +1,428 @@
+ * should call this function yourself or ensure that whatever
+ * initialized OpenSSL cleans it up too.
+ */
+void cleanupOpenSSL();
+
+/**
+ * OpenSSL implementation for SSL socket interface.
+ */
+class TSSLSocket : public TSocket {
+public:
+  ~TSSLSocket() override;
+/**
+ * TTransport interface.
+ */
+#if 0
+  bool isOpen() const override;
+  bool peek() override;
+  void open() override;
+  void close() override;
+  bool hasPendingDataToRead() override;
+  uint32_t read(uint8_t* buf, uint32_t len) override;
+  void write(const uint8_t* buf, uint32_t len) override;
+  uint32_t write_partial(const uint8_t* buf, uint32_t len) override;
+  void flush() override;
+#endif
+  void open() override;
+  /**
+   * Set whether to use client or server side SSL handshake protocol.
+   *
+   * @param flag  Use server side handshake protocol if true.
+   */
+  void server(bool flag) { server_ = flag; }
+  /**
+   * Determine whether the SSL socket is server or client mode.
+   */
+  bool server() const { return server_; }
+  /**
+   * Set AccessManager.
+   *
+   * @param manager  Instance of AccessManager
+   */
+  virtual void access(std::shared_ptr<AccessManager> manager) { access_ = manager; }
+  /**
+   * Set eventSafe flag if libevent is used.
+   */
+  void setLibeventSafe() { eventSafe_ = true; }
+  /**
+   * Determines whether SSL Socket is libevent safe or not.
+   */
+  bool isLibeventSafe() const { return eventSafe_; }
+
+  void authenticate(bool required);
+
+protected:
+  /**
+   * Constructor.
+   */
+  TSSLSocket(std::shared_ptr<TConfiguration> config = nullptr);
+  /**
+   * Constructor with an interrupt signal.
+   */
+  TSSLSocket(std::shared_ptr<THRIFT_SOCKET> interruptListener,
+             std::shared_ptr<TConfiguration> config = nullptr);
+  /**
+   * Constructor, create an instance of TSSLSocket given an existing socket.
+   *
+   * @param socket An existing socket
+   */
+  TSSLSocket(THRIFT_SOCKET socket, std::shared_ptr<TConfiguration> config = nullptr);
+  /**
+   * Constructor, create an instance of TSSLSocket given an existing socket that can be interrupted.
+   *
+   * @param socket An existing socket
+   */
+  TSSLSocket(THRIFT_SOCKET socket,
+             std::shared_ptr<THRIFT_SOCKET> interruptListener,
+             std::shared_ptr<TConfiguration> config = nullptr);
+  /**
+   * Constructor.
+   *
+   * @param host  Remote host name
+   * @param port  Remote port number
+   */
+  TSSLSocket(std::string host, int port, std::shared_ptr<TConfiguration> config = nullptr);
+  /**
+   * Constructor with an interrupt signal.
+   *
+   * @param host  Remote host name
+   * @param port  Remote port number
+   */
+  TSSLSocket(std::string host,
+             int port,
+             std::shared_ptr<THRIFT_SOCKET> interruptListener,
+             std::shared_ptr<TConfiguration> config = nullptr);
+
+  void setVerifyMode(int mode);
+
+  void openSecConnection(struct addrinfo* res);
+
+  bool server_;
+
```

Comment:
Also note, that this was removed, which might actually be kind of useful for us.
```
-  std::shared_ptr<SSLContext> ctx_;
```